compiler/lib-wasm/wasm_source_map.ml

```diff
@@ -0,0 +1,174 @@
+open Stdlib
+
+type resize_data =
+  { mutable i : int
+  ; mutable pos : int array
+  ; mutable delta : int array
+  }
+
+type t = Yojson.Raw.t
+
+type input = Vlq64.input =
+  { string : string
+  ; mutable pos : int
+  ; len : int
+  }
+
+let rec next' src mappings pos =
+  pos < src.len
+  &&
+  match mappings.[pos] with
+  | ',' ->
+      src.pos <- pos + 1;
+      true
+  | _ -> next' src mappings (pos + 1)
+
+let next src = next' src src.string src.pos
+
+let flush buf src start pos =
+  if start < pos then Buffer.add_substring buf src.string start (pos - start)
+
+let rec resize_rec buf start src resize_data i col0 delta0 col =
+  let pos = src.pos in
+  let delta = Vlq64.decode src in
+  let col = col + delta in
+  if col < col0
+  then
+    if next src
+    then resize_rec buf start src resize_data i col0 delta0 col
+    else flush buf src start (String.length src.string)
+  else
+    let delta = delta + delta0 in
+    adjust buf start src resize_data i col delta pos
+
+and adjust buf start src (resize_data : resize_data) i col delta pos =
+  assert (delta > 0);
+  if i < resize_data.i
+  then
+    let col0 = resize_data.pos.(i) in
+    let delta0 = resize_data.delta.(i) in
+    if col < col0
+    then (
+      flush buf src start pos;
+      Vlq64.encode buf delta;
+      let start = src.pos in
+      if next src
+      then resize_rec buf start src resize_data (i + 1) col0 delta0 col
+      else flush buf src start (String.length src.string))
+    else
+      let delta = delta + delta0 in
+      adjust buf start src resize_data (i + 1) col delta pos
+  else (
+    flush buf src start pos;
+    Vlq64.encode buf delta;
+    let start = src.pos in
+    flush buf src start (String.length src.string))
+
+let resize_mappings (resize_data : resize_data) mappings =
+  if String.equal mappings "" || resize_data.i = 0
+  then mappings
+  else
+    let col0 = resize_data.pos.(0) in
+    let delta0 = resize_data.delta.(0) in
+    let buf = Buffer.create (String.length mappings) in
+    resize_rec
+      buf
+      0
+      { Vlq64.string = mappings; pos = 0; len = String.length mappings }
+      resize_data
+      1
+      col0
+      delta0
+      0;
+    Buffer.contents buf
+
+let resize resize_data sm =
```

Comment:
What's this logic about ? Can we use Index sourcemap instead ?

Reply:
It's a bit as if we were reformatting some JavaScript code. We could implement the corresponding source map update by splitting the input source map into an index containing a large number of output source maps, for each unchanged piece of code. But it makes a lot more sense to keep a single standard source map and adjust the position of the mappings.

Reply:
> For this, one needs to renumber the module components

This is the part I don't really understand, can you point me to where the renumbering of components happens ?

Reply:
We are building a single Wasm module by linking together several Wasm modules.
There is a function `scanner` in `wasm_link.ml` which scan the Wasm binary code and renumber the components.
https://github.com/ocsigen/js_of_ocaml/blob/dcd52fc075934d9f7b800b60df09f2b1d58c8bbd/compiler/lib-wasm/wasm_link.ml#L894
The renumbering is performed by functions `rewrite` and `rewrite_signed`, which read the index and write the updated index, keeping track of code size changes.
https://github.com/ocsigen/js_of_ocaml/blob/dcd52fc075934d9f7b800b60df09f2b1d58c8bbd/compiler/lib-wasm/wasm_link.ml#L950